src/cascadia/TerminalApp/CommandPalette.cpp

```diff
@@ -0,0 +1,396 @@
+    //   for which mode we should enter in.
+    void CommandPalette::ToggleVisibility()
+    {
+        const bool isVisible = Visibility() == Visibility::Visible;
+        if (!isVisible)
+        {
+            // Become visible
+            Visibility(Visibility::Visible);
+            _SearchBox().Focus(FocusState::Programmatic);
+            _FilteredActionsView().SelectedIndex(0);
+        }
+        else
+        {
+            // Raise an event to return control to the Terminal.
+            _close();
+        }
+    }
+
+    // Method Description:
+    // - Moves the focus up or down the list of commands. If we're at the top,
+    //   we'll loop around to the bottom, and vice-versa.
+    // Arguments:
+    // - moveDown: if true, we're attempting to move to the next item in the
+    //   list. Otherwise, we're attempting to move to the previous.
+    // Return Value:
+    // - <none>
+    void CommandPalette::_selectNextItem(const bool moveDown)
+    {
+        const auto selected = _FilteredActionsView().SelectedIndex();
+        const int numItems = ::base::saturated_cast<int>(_FilteredActionsView().Items().Size());
+        // Wraparound math. By adding numItems and then calculating modulo numItems,
+        // we clamp the values to the range [0, numItems) while still supporting moving
+        // upward from 0 to numItems - 1.
+        const auto newIndex = ((numItems + selected + (moveDown ? 1 : -1)) % numItems);
+        _FilteredActionsView().SelectedIndex(newIndex);
+        _FilteredActionsView().ScrollIntoView(_FilteredActionsView().SelectedItem());
+    }
+
+    // Method Description:
+    // - Process keystrokes in the input box. This is used for moving focus up
+    //   and down the list of commands in Action mode, and for executing
+    //   commands in both Action mode and Commandline mode.
+    // Arguments:
+    // - e: the KeyRoutedEventArgs containing info about the keystroke.
+    // Return Value:
+    // - <none>
+    void CommandPalette::_keyDownHandler(IInspectable const& /*sender*/,
+                                         Windows::UI::Xaml::Input::KeyRoutedEventArgs const& e)
+    {
+        auto key = e.OriginalKey();
+
+        if (key == VirtualKey::Up)
+        {
+            // Action Mode: Move focus to the next item in the list.
+            _selectNextItem(false);
+            e.Handled(true);
+        }
+        else if (key == VirtualKey::Down)
+        {
+            // Action Mode: Move focus to the previous item in the list.
+            _selectNextItem(true);
+            e.Handled(true);
+        }
+        else if (key == VirtualKey::Enter)
+        {
+            // Action Mode: Dispatch the action of the selected command.
+
+            if (const auto selectedItem = _FilteredActionsView().SelectedItem())
+            {
+                if (const auto data = selectedItem.try_as<Command>())
+                {
+                    const auto actionAndArgs = data.Action();
```

Comment:
So there'll already be "nested commands", where you can hit enter on "New Tab...", and it'll switch the list to display "profile 1, profile 2, profile 3, etc...". That kind of model could also work for the tab switcher. It'd be easy to hook up the left/right arrows to do that as well. 

Displaying the nested commands w/in the parent `ListViewItem` might get harder, but it could _probably_ be done